ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+we'll use the method `#execute` as the primary entrypoint for executing user code.
+
+```ruby
+class MyActivity < Temporal::Activity
+  ManuallyCancelled = Class.new(Temporal::Error)
+
+  def execute(arg_1, arg_2: nil)
+    loop do
+      sleep 1
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::ActivityCancelled
+    raise ManuallyCancelled, 'activity got cancelled'
+  end
+end
+```
+
+Notes:
+
+- `activity` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- Errors defined within the activity class will be namespaces `MyActivity::ManuallyCancelled`,
+  which is a very useful property
+
+```ruby
+class Temporal::Activity::Context
+  # Access to a pre-configured logger
+  def logger() -> Temporal::Logger
+
+  # Heartbeat will raise if a cancellation was requested
+  def heartbeat(*details) -> void
+
+  # Provides access to extra details from PollActivityTaskQueueResponse
+  def info() -> Temporal::Info::Activity
+end
+```
+
+
+## Payload converters
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to facilitate conversion between SDK inputs and payloads. Basically there will be two extension
+points on the `Temporal::Configuration` object allowing to add either a DataConverter or a Codec.
+
+The main difference is that a DataConverter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the later is for
+modifying a binary that is transmitted over the wire.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7933')
+
+config.add_data_converter(MyProtobufConverter.new)
+config.add_data_converter(MyThriftConverter.new)
+
+config.add_codec(EncryptionCodec.new(KEY_ID))
+config.add_codec(CompressionCodec.new)
+```
+
+A data converter is expected to have this interface:
+
+```ruby
+class Temporal::DataConverter::Base
+  # Returns mime-type of the payload
+  def encoding() -> String
+
+  # Takes any Ruby object and returns a Payload object or nil if unable to convert
+  def to_payload(data: Any) -> Temporal::Payload
+
+  # Takes a Payload object and returns a Ruby object. Only called on a matching encoding
+  def from_payload(payload: Temporal::Payload) -> Any
+end
+```
+
+And a code is expected to have this interface:
+
+```ruby
+class Temporal::Codec::Base
+  # Takes Payloads object and returns a modified Payloads object
+  def encode(payloads: Temporal::Payloads) -> Temporal::Payloads
```

Comment:
Should be from `Payload[]` to `Payload[]`.
Core doesn't use the `Payloads` wrapper.

Reply:
The main reason for using `Payloads` is I was told that Temporal might add additional fields to it at some point, but I'll change it to `Payload[]`, no problem 👍